rasa/utils/tensorflow/layers.py

```diff
@@ -1137,11 +1146,36 @@ def _compute_softmax_loss(
                 [softmax_logits, sim_neg_ii, sim_neg_ll], axis=-1
             )
         # create label_ids for softmax
-        softmax_label_ids = tf.zeros_like(softmax_logits[..., 0], tf.int32)
-        softmax_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
-            labels=softmax_label_ids, logits=softmax_logits
+        # label_ids = tf.concat(
+        #     [
+        #         tf.ones_like(softmax_logits[..., :1], tf.float32),
+        #         tf.zeros_like(softmax_logits[..., 1:], tf.float32),
+        #     ],
+        #     axis=-1,
+        # )
+
+        # for numerical stability of softmax,
+        # see https://stackoverflow.com/a/49212689/3001665
+        input_logits = tf.nn.softmax(
+            softmax_logits - tf.math.reduce_max(softmax_logits, axis=-1, keepdims=True)
```

Comment:
shifting the logits should only be necessary if you implemented the softmax by hand

Reply:
This was an additional harmless safeguard because `tf.nn.softmax` is implemented as a C++ function in tensorflow which is not visible in the TF source code right now unless we install it from source using `bazel` (the python version uses it through `gen_nn_ops` module which contains all the compiled C++ code in python). I didn't want to get caught up in doing that right now, can do once we are sure that making these changes is useful.